jme3-core/src/main/java/com/jme3/renderer/opengl/GLRenderer.java

```diff
@@ -1379,7 +1423,17 @@ public void updateShaderSourceData(ShaderSource source) {
             // Error P0001: Extension directive must occur before any non-preprocessor tokens
             int idx = stringBuf.lastIndexOf("#extension");
             idx = stringBuf.indexOf("\n", idx);
-            stringBuf.insert(idx + 1, "precision mediump float;\n");
+
+            if(version>=310) {
+                stringBuf.insert(idx + 1, "precision highp sampler2DMS;\n");
+            }
+            if(version>=300) {
+                stringBuf.insert(idx + 1, "precision highp sampler2DArray;\n");
+                stringBuf.insert(idx + 1, "precision highp sampler2DShadow;\n");
+                stringBuf.insert(idx + 1, "precision highp sampler3D;\n");
+                stringBuf.insert(idx + 1, "precision highp sampler2D;\n");
+            }
+            stringBuf.insert(idx + 1, "precision highp float;\n");
```

Comment:
Is it a good idea to force this on the developer?
I know the code already forced precision mediump float; , but this looks like bad practice to me.

Reply:
So you mean removing all these inserts and making the game/shader developer? This will lead to plenty of current jme3 shaders not to compile on gles at all.

Reply:
Could it be implemented in GLSLCompat? So that devs can leave it out for custom shaders.

Reply:
Not really, precision directive must be after last #extension line. All shaders include GLSLCompat, then other requirements (that could have extension directives) and then the shader code, so having precision defined in GLSLCompat will make compilation fail

Reply:
Isn't the shader loader supposed to move the #extension directives to the top?

Reply:
Yes, you're right and should work. I'll change and test it and get back to you

Reply:
I've tried changed it and some shaders included by default in jme3 like Common/MatDefs/Terrain/TerrainLighting.frag don't include GLSLCompat.glsllib. We could either leave it as it's now or review all shaders to include GLSLCompat as you think is a better option :thinking:

Reply:
Yeah, not a problem, we can add the GLSLCompat with a future PR,